Fix `email_on_failure` with `render_template_as_native_obj`

jedcunningham opened the pull request:
Co-authored-by: andyhuang <andyhuang@mirrormedia.mg>

Closes: #22152
Alternative of #22218, refactored to only change behavior when generating emails and including test coverage.

cc @andyfcx

jedcunningham committed:
Fix `email_on_failure` with `render_template_as_native_obj`

Co-authored-by: andyhuang <andyhuang@mirrormedia.mg>

uranusjr commented on airflow/models/abstractoperator.py:
```diff
@@ -106,7 +106,7 @@ def dag_id(self) -> str:
         return self.task_id
 
-    def get_template_env(self) -> "jinja2.Environment":
+    def get_template_env(self, force_sandboxed: bool = False) -> "jinja2.Environment":
```
```suggestion
    def get_template_env(self, *, force_sandboxed: bool = False) -> "jinja2.Environment":
```

Let’s make this keyword-only for explicitness.

uranusjr commented on airflow/models/dag.py:
```diff
@@ -1260,7 +1260,7 @@ def resolve_template_files(self):
             t.resolve_template_files()
 
-    def get_template_env(self) -> jinja2.Environment:
+    def get_template_env(self, force_sandboxed: bool = False) -> jinja2.Environment:
```
```suggestion
    def get_template_env(self, *, force_sandboxed: bool = False) -> jinja2.Environment:
```

uranusjr reviewed (approved):
Some nitpicks otherwise lgtm.

jedcunningham committed:
Apply suggestions from code review

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

dstandish commented on tests/models/test_dag.py:
```diff
@@ -473,6 +474,18 @@ def test_template_undefined(self):
 
+    @parameterized.expand(
+        [
+            (False, False, SandboxedEnvironment),
```
```suggestion
            (False, True, SandboxedEnvironment),
            (False, False, SandboxedEnvironment),
```

dstandish commented on airflow/models/taskinstance.py:
```diff
@@ -2196,7 +2196,7 @@ def get_email_subject_content(self, exception: BaseException) -> Tuple[str, str,
 
         else:
-            jinja_env = self.task.get_template_env()
+            jinja_env = self.task.get_template_env(force_sandboxed=True)
```
i think you may need to update this operator to handle kwargs
https://github.com/apache/airflow/blob/be0a4e413117c5ceb0a245402cdfabe1a3d82ace/airflow/providers/microsoft/psrp/operators/psrp.py#L158

also, this illustrates a backcompat issue if a user implemented this method (since it's a signature change).

if we wanted to be real careful we could inspect the signature before calling....

wdyt?

ashb replied:
Oh ugh that's going to make this approach hard/basically impossible. Nice catch @dstandish

jedcunningham replied:
I think I lean toward this being "internal" to the task, but I'm not sure how others view it.

We could inspect, but emails would still fail even then. I wonder if we should instead release a new psrp provider that has a minimum core of 2.3.0. I guess the most 'flexible', ignoring the email issue, is to inspect in both so new providers still work on old core and vice versa.

ashb replied:
We could use `dag.get_template_env(force_sandbox)` instead of `self` in side the email rendering?

jedcunningham replied:
Oh, good idea, it uses `dag` anyways! Let me try that.

jedcunningham replied:
Yeah, that works 🍺

blag reviewed (commented):
Besides @dstandish's questions, LGTM.

jedcunningham committed:
Update tests/models/test_dag.py

Co-authored-by: Daniel Standish <15932138+dstandish@users.noreply.github.com>

jedcunningham committed:
Use DAGs `get_template_env` instead of the tasks

jedcunningham commented on airflow/models/abstractoperator.py:
```diff
@@ -106,7 +106,7 @@ def dag_id(self) -> str:
         return self.task_id
 
-    def get_template_env(self) -> "jinja2.Environment":
+    def get_template_env(self, *, force_sandboxed: bool = False) -> "jinja2.Environment":
```
Think we should remove support for `force_sandboxed` here so we arent changing the signature of the method psrp is overriding?

ashb replied:
Doesn't do any harm since its an optional arg, but for the sake of a smaller diff, yes we could remove it.

uranusjr replied:
+1 to removing it since it’d make more obvious why we call the DAG’s `get_template_env` instead (so we can pass this option). It’d be nice to have a comment there describing why we don’t call the task’s `get_template_env` there.

uranusjr replied:
I pushed a commit for this (plus a small tweak for Mypy).

uranusjr committed:
Fix DAG access error

Also added comments on why we use the DAG level implementation instead
of the task level one.
